xfrd.c

```diff
@@ -2383,78 +2525,255 @@ find_same_master_notify(xfrd_zone_type* zone, int acl_num_nfy)
+		       xfr->state == xfrd_xfr_new);
+		if(xfr == zone->latest_xfr) {
+			zone->soa_xfr_acquired = xfr->acquired;
+			zone->soa_xfr = xfr->soa;
+		}
+		xfrd_schedule_zone_xfrs(zone, xfr, committed);
+		return; /* nothing to update yet, continue */
+	}
+
+	assert(committed);
+	assert(xfr && (xfr->state == xfrd_xfr_ok ||
+	               xfr->state == xfrd_xfr_new));
+
+	/* transfer committed to memory */
+	log_msg(LOG_INFO, "zone %s serial %u is updated to %u",
+		zone->apex_str,
+		ntohl(zone->soa_nsd.serial),
+		ntohl(xfr->soa.serial));
+	zone->soa_nsd = xfr->soa;
+	zone->soa_nsd_acquired = xfr->acquired;
+	xfrd->write_zonefile_needed = 1;
+
+	/* discard committed transfers */
+	while(xfr) {
+		prev_xfr = xfr->prev;
+		xfrd_discard_zone_xfr(zone, xfr);
+		xfr = prev_xfr;
+	}
+
+	if((!zone->latest_xfr) ||
+	   (!zone->latest_xfr->prev || !zone->latest_xfr->acquired))
+	{
+		zone->soa_xfr_acquired = 0;
+		memset(&zone->soa_xfr, 0, sizeof(xfrd_soa_type));
+	}
+
+	xfrd_schedule_zone_xfrs(zone, zone->latest_xfr, committed);
+	xfrd_send_notify(xfrd->notify_zones, zone->apex, &zone->soa_nsd);
 }
 
+/* zone transfers must be processed after all soa information returned by nsd
+   has been processed, it is unknown whether difffiles for successful transfers
+   can be discarded otherwise. */
 void
-xfrd_prepare_zones_for_reload()
+xfrd_process_xfrs(int committed)
 {
-	xfrd_zone_type* zone;
+	xfrd_zone_type *zone;
+
 	RBTREE_FOR(zone, xfrd_zone_type*, xfrd->zones)
 	{
-		/* zone has a disk soa, and no nsd soa or a different nsd soa */
-		if(zone->soa_disk_acquired != 0 &&
-			(zone->soa_nsd_acquired == 0 ||
-			zone->soa_disk.serial != zone->soa_nsd.serial))
+		time_t elapsed;
+
+		/* reloads are scheduled by xfrd_schedule_zone_xfrs if any
+		   transfers are added to the tasklist */
+		xfrd_process_zone_xfrs(zone, committed);
+
+		if(!committed) {
+			continue; /* continue if nothing has been committed */
+		}
+
+		if((zone->soa_notified_acquired != 0 &&
+		    compare_serial(ntohl(zone->soa_notified.serial),
+		                   ntohl(zone->soa_nsd.serial)) <= 0) ||
+		   (zone->soa_notified_acquired < zone->soa_nsd_acquired))
 		{
-			if(zone->soa_disk_acquired == xfrd_time()) {
-				/* antedate by one second.
-				 * this makes sure that the zone time is before
-				 * reload, so that check_failed_zones() is
-				 * certain of the result.
-				 */
-				zone->soa_disk_acquired--;
+			zone->soa_notified_acquired = 0;
+			memset(&zone->soa_notified, 0, sizeof(xfrd_soa_type));
+		}
+
+		/* earlier versions of NSD checked if a transfer was completed
+		   before the reload and concluded that the update must have
+		   failed if the transfer was still around. that shouldn't be
+		   necessary anymore since transfers are now checked
+		   individually. still, the logic is off if a completed
+		   transfer is around after the reload */
+		//assert(zone->soa_disk_acquired &&
+		//       zone->soa_disk_acquired <= xfrd->reload_cmd_last_sent);
```

Comment:
Yes, and I would like an assertion for that.  Also, the `//` operand is imported from C++ and not in early C standards, so I think we should not use that, that improves portability.

Reply:
Reinstated.